In [1]:
import networkx as nx
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import powerlaw
warnings.filterwarnings('ignore')
%matplotlib inline

Watts-Strogatz (small world) graphs: model for undirected networks with large clustering coefficient and short distances, as observed in real networks. The model needs three values: N nodes, K number of neighbours each link is linked to, and p probability of a link being rewired randomly to a different node (not allowing self-links).

It should be remarked immediately that our network (treated as an undirected network) has a not-so-high clustering coefficient: according to what was calculated by Chiara M., the avg. clustering coefficient should be 0.165.

<p>**Nota**: provando a calcolare lo stesso numero con gephi, ho notato che per avere l'avg. clustering coefficient è necessario usare un plugin Clustering coefficient, selezionando Basic come opzione. Dopo che ha finito i calcoli (è molto lento, e può dare l'impressione che il programma sia bloccato), esce una schermata con tutti i C per ogni nodo, ed in fondo c'è il C che coincide con l'avg. C del network. Questo risultato coincide con i risultati di networkx (con test a caso, anche il C locale coincideva, almeno fino alla 4a-5a cifra).</p>
<p>Tuttavia, con l'Avg. C presente nel programma di default (sotto la sezione Network), il risultato è 0,311. Metodo di calcolo diverso? Sto provando ad informarmi. L'opzione Triangle method del plugin per il C suppongo sia invece per il calcolo del global clustering coefficient? Il programma dà risultato 0,0279.</p>

Treating our network as if it was an undirected network, the average degree of the nodes (as calculated by Gephi, and the number is confirmed by the analysis by Chiara M.) should be 8.404. I've thus made two models, one with k=8 and one with k=9, to have lower and upper bounds (chosing 10 instead of 9 because the network x WS function to create networks creates k-1 edges in case k is an odd number).<br />
In addition, according to the calculations on Gephi, the avg. path length (a.k.a. average length of the shortest path) is estimated at 3.736, with a network diameter (max. distance) of 10.

## Regular lattice
The WS model as a regular lattice (p=0):

In [27]:
#24819 nodes
#104284 edges

# network generators 2 (small-world)
wsLow = nx.watts_strogatz_graph(24819, k=8, p=0)
wsUp = nx.watts_strogatz_graph(24819, k=10, p=0)
#nx.draw(ws)
print("Links of k=8: ", wsLow.number_of_edges())
print("Links of k=9: ", wsUp.number_of_edges())
print("Our network has 24819 nodes and 104284 links.")

Links of k=8:  99276
Links of k=9:  124095
Our network has 24819 nodes and 104284 links.


In [91]:
print(f"Global clustering (lower bound): {nx.average_clustering(wsLow)}")
print(f"Global clustering (upper bound): {nx.average_clustering(wsUp)}")

Global clustering (lower bound): 0.6428571428569719
Global clustering (upper bound): 0.6666666666664383


The clustering coefficient is significantly higher than the one observed in our network.

In [77]:
print(f"Average shortest path (lower bound): {nx.average_shortest_path_length(wsLow)}")
print(f"Average shortest path (upper bound): {nx.average_shortest_path_length(wsUp)}")

Average shortest path (low bound): 1551.6250302200017
Average shortest path (upper bound): 1241.4000322346683


In [9]:
# various metrics
connectL = nx.is_connected(wsLow)
print("L=8 is strongly connected: ", connectL)
connectU = nx.is_connected(wsUp)
print("L=9 is strongly connected: ", connectU)

L=8 is strongly connected:  True
L=9 is strongly connected:  True


In [10]:
densityL = nx.density(wsLow)
densityU = nx.density(wsUp)
print("L=8 density: ", densityL)
print("L=9 density: ", densityU)

L=8 density:  0.0003223466838584898
L=9 density:  0.00040293335482311226


## Random network
Random network (p=1):

In [29]:
# network generators 2 (small-world)
wsLow = nx.watts_strogatz_graph(24819, k=8, p=1)
wsUp = nx.watts_strogatz_graph(24819, k=10, p=1)
#nx.draw(ws)

In [93]:
print(f"Global clustering (lower bound): {nx.average_clustering(wsLow)}")
print(f"Global clustering (upper bound): {nx.average_clustering(wsUp)}")

Global clustering (lower bound): 0.0002601078450226278
Global clustering (upper bound): 0.00036301361439011347


The clustering coefficient is significantly smaller than the one observed in our network. The average shortest path length, however, is much closer to our network's:

In [86]:
print(f"Average shortest path (lower bound): {nx.average_shortest_path_length(wsLow)}")
print(f"Average shortest path (upper bound): {nx.average_shortest_path_length(wsUp)}")

Average shortest path (lower bound): 5.173419489735226
Average shortest path (upper bound): 4.685677318533544


In [15]:
# various metrics
connectL = nx.is_connected(wsLow)
print("L=8 is strongly connected: ", connectL)
connectU = nx.is_connected(wsUp)
print("L=9 is strongly connected: ", connectU)

L=8 is strongly connected:  True
L=9 is strongly connected:  True


In [16]:
densityL = nx.density(wsLow)
densityU = nx.density(wsUp)
print("L=8 density: ", densityL)
print("L=9 density: ", densityU)

L=8 density:  0.0003223466838584898
L=9 density:  0.00040293335482311226


## Small world regime
The small world regime consists in a network with p's value such as 0.001 < p < 0.1.

With a p of 0.5, we still obtain a global clustering coefficient and avg. path length that are still significantly higher than the ones observed in our network:

In [94]:
# network generators 2 (small-world)
wsLow = nx.watts_strogatz_graph(24819, k=8, p=0.05)
wsUp = nx.watts_strogatz_graph(24819, k=10, p=0.05)
#nx.draw(ws)

In [95]:
print(f"Global clustering (lower bound): {nx.average_clustering(wsLow)}")
print(f"Global clustering (upper bound): {nx.average_clustering(wsUp)}")

Global clustering (lower bound): 0.5538923974074691
Global clustering (upper bound): 0.5725695690702096


In [96]:
print(f"Average shortest path (lower bound): {nx.average_shortest_path_length(wsLow)}")
print(f"Average shortest path (upper bound): {nx.average_shortest_path_length(wsUp)}")

Average shortest path (lower bound): 9.750140161355368
Average shortest path (upper bound): 8.24615103672127


A value of p close to 0.4 seems to give the avg. clustering coefficient closest to our network's.

In [20]:
# network generators 3 prova mia
wsLow = nx.watts_strogatz_graph(24819, k=8, p=0.37)
wsUp = nx.watts_strogatz_graph(24819, k=10, p=0.37)
#nx.draw(ws)

In [127]:
print(f"Global clustering (lower bound): {nx.average_clustering(wsLow)}")
print(f"Global clustering (upper bound): {nx.average_clustering(wsUp)}")

Global clustering (lower bound): 0.16438924439649966
Global clustering (upper bound): 0.16951747257997285


In [128]:
print(f"Average shortest path (lower bound): {nx.average_shortest_path_length(wsLow)}")
print(f"Average shortest path (upper bound): {nx.average_shortest_path_length(wsUp)}")

Average shortest path (lower bound): 5.687665740009242
Average shortest path (upper bound): 5.132950940991358


The average shortest path length is also much closer to our network's (albeit not as low).

In [22]:
# various metrics
connectL = nx.is_connected(wsLow)
print("L=8 is strongly connected: ", connectL)
connectU = nx.is_connected(wsUp)
print("L=9 is strongly connected: ", connectU)

L=8 is strongly connected:  True
L=9 is strongly connected:  True


In [23]:
densityL = nx.density(wsLow)
densityU = nx.density(wsUp)
print("L=8 density: ", densityL)
print("L=9 density: ", densityU)

L=8 density:  0.0003223466838584898
L=9 density:  0.00040293335482311226


+ connected
+ avg clustering coefficient (che è il valore che ho calcolato)
+ density (calcolata, è uguale per tutti ovviamente, magari cancello dopo la prima visto che è ovvio)
+ i vari k sono come da modello per tutti, rimangono costanti
+ avg shortest path (che è il valore che ho calcolato)
+ distribution (deriva dalla teoria)
+ lattices: dirac delta function
+ others: poisson

<style>
    th, td, tr {
        border: 1px solid black !important;

</style>
<table>
    <tr>
        <th></th>
        <th colspan="2">Lattice</th>
        <th colspan="2">Random network</th>
        <th colspan="2">Small world regime</th>
    </tr>
    <tr>
        <th></th>
        <th>K=8</th>
        <th>K=9</th>
        <th>K=8</th>
        <th>K=9</th>
        <th>K=8</th>
        <th>K=9</th>
    </tr>
    <tr>
        <td>Connectedness<td>
        <td colspan="6">Connected</td>
    </tr>
    <tr>
        <td>Avg. clustering coefficient</td>
        <td>0.6428</td>
        <td>0.6667</td>
        <td>0.0003</td>
        <td>0.0004</td>
        <td>0.1644</td>
        <td>0.1695</td>
    </tr>
    <tr>
        <td>Density</td>
        <td>0.0003</td>
        <td>0.0004</td>
        <td>0.0003</td>
        <td>0.0004</td>
        <td>0.0003</td>
        <td>0.0004</td>
    </tr>
    <tr>
        <td>K (min, max, avg)</td>
        <td>8</td>
        <td>9</td>
        <td>8</td>
        <td>9</td>
        <td>8</td>
        <td>9</td>
    </tr>
    <tr>
        <td>Avg. shortest path</td>
        <td>1551.625</td>
        <td>1241.400</td>
        <td>5.173</td>
        <td>4.686</td>
        <td>5.688</td>
        <td>5.133</td>
    </tr>
    <tr>
        <td>Distribution</td>
        <td colspan="2">Dirac delta function</td>
        <td colspan="2">Poisson</td>
        <td colspan="2">Poisson</td>
    </tr>
</table>
